xla/backends/gpu/runtime/dynamic_slice_thunk.h

```diff
@@ -47,8 +47,44 @@ class DynamicSliceThunk : public Thunk {
  public:
   // Dynamic slice offset can be either: (1) a statically known constant value
   // or (2) a truly dynamic offset that is computed on device and have to be
-  // transferred to host.
-  using Offset = std::variant<int64_t, BufferAllocation::Slice>;
+  // transferred to host or (3) a temporary HloModule that computes the offset
+  // with a single induction variable as the input.
+  using Offset = std::variant<int64_t, BufferAllocation::Slice, HloModule*>;
+
+  struct OffsetAsFunctionOfIndvarModulesMetadata {
+    // These two modules help keep track of the induction variable. The module
+    // `indvar_init_` is a module with prototype `() -> integer[]`. It takes
+    // no input, and returns the initial value of the induction variable. The
+    // module `indvar_update_` is a module with prototype `(integer[]) ->
+    // integer[]`. It takes the current value of the induction variable, and
+    // returns the next value of the induction variable.
+    std::unique_ptr<HloModule> indvar_init_, indvar_update_;
+
+    // Extracted HloModules for computing dynamic offsets. The modules are
+    // not used here, this is solely for keeping the modules alive and maintain
+    // their ownership with the thunk while their raw pointers would be used
+    // during execution from the `offsets_` vector. These modules are with
+    // signature `(integer[]) -> integer[]`, where the input is the current
+    // value of the loop induction variable, and the output is the offset value
+    // for that iteration.
+    std::vector<std::unique_ptr<HloModule>> extracted_offset_modules;
+
+    OffsetAsFunctionOfIndvarModulesMetadata(
+        std::unique_ptr<HloModule> indvar_init,
+        std::unique_ptr<HloModule> indvar_update,
+        std::vector<std::unique_ptr<HloModule>> extracted_offset_modules)
+        : indvar_init_(std::move(indvar_init)),
+          indvar_update_(std::move(indvar_update)),
+          extracted_offset_modules(std::move(extracted_offset_modules)) {
+      CHECK(indvar_init_ != nullptr && indvar_update_ != nullptr);
+      CHECK(indvar_init_->entry_computation()->num_parameters() == 0)
```

Comment:
nit: make a conditional that checks the output signatures as well?

Note: I'll LGTM to try to import the PR and avoid the back and forth since this is a nit comment, but would appreciate if you could make a tiny PR adding the checks :)